### PR TITLE
[CLI] Add a subcommand for getting information about a keyboard

### PR DESCRIPTION
**Pull request by skullydazed**
This adds a new command, `qmk info`, that parses info.json and displays some information about the keyboard. You can specify the keyboard with -kb, by setting `user.keyboard`, or with the `automagic_keyboard` functionality.

You can also render the layouts in various ways:

* `-l` Show all the layouts with info.json labels
* `-m` Show all the layouts with matrix positions
* `-km <keymap>` Show the layers for a JSON keymap

## Types of Changes

- [x] Core
- [x] New feature

## Checklist

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by skullydazed** [comment]
## Showing the command in action

### Basic info in text format

Screen Shot 2020-04-07 at 13 11 16

### Matrix and Keymap printing

Screen Shot 2020-04-07 at 13 08 29

### Layout printing

Screen Shot 2020-04-05 at 15 50 36

File: lib/python/qmk/constants.py [hunk 1]
```diff
@@ -7,3 +7,9 @@
 
 # This is the number of directories under `qmk_firmware/keyboards` that will be traversed. This is currently a limitation of our make system.
 MAX_KEYBOARD_SUBFOLDERS = 5
+
+# Supported processor types
+ARM_PROCESSORS = 'cortex-m0', 'cortex-m0plus', 'cortex-m3', 'cortex-m4', 'MKL26Z64', 'MK20DX128', 'MK20DX256', 'STM32F042', 'STM32F072', 'STM32F103', 'STM32F303'
+AVR_PROCESSORS = 'at90usb1286', 'at90usb646', 'atmega16u2', 'atmega328p', 'atmega32a', 'atmega32u2', 'atmega32u4', None
+ALL_PROCESSORS = ARM_PROCESSORS + AVR_PROCESSORS
+VUSB_PROCESSORS = 'atmega328p', 'atmega32a'
```
Comment (Erovia) [inline lib/python/qmk/constants.py:15]: Just a mental note to myself, that this well need to merged somehow with
https://github.com/noroadsleft/qmk_firmware/blob/feature/arm_template/lib/python/qmk/constants.py

**Comment by skullydazed** [comment]
CI failure is unrelated, same error occurs on master for `handwired/onekey/pytest:rgb`.

**Comment by skullydazed** [comment]
This should be ready to go now.

File: lib/python/qmk/makefile.py [hunk 2]
```diff
@@ -2,8 +2,6 @@
 """
 from pathlib import Path
 
-from qmk.errors import NoSuchKeyboardError
-
 
 def parse_rules_mk_file(file, rules_mk=None):
```
[the remark this answers is not in the document]
Comment (skullydazed) [reply lib/python/qmk/makefile.py:6]: The reason I separated them is that generally things in `qmk.keyboard` are things that subcommands should use while things in `qmk.makefile` are not.

File: lib/python/qmk/path.py [hunk 3]
```diff
@@ -4,7 +4,7 @@
 import os
 from pathlib import Path
 
-from qmk.constants import QMK_FIRMWARE, MAX_KEYBOARD_SUBFOLDERS
+from qmk.constants import MAX_KEYBOARD_SUBFOLDERS, QMK_FIRMWARE
```
[the remark this answers is not in the document]
Comment (skullydazed) [reply lib/python/qmk/path.py:7]: Alphabatizing.

File: lib/python/qmk/keymap.py [hunk 4]
```diff
@@ -95,32 +95,71 @@ def write(keyboard, keymap, layout, layers):
...
+        if checked_dirs:
+            checked_dirs = '/'.join((checked_dirs, dir))
+        else:
+            checked_dirs = dir
+
+        keymap_dir = Path('keyboards') / checked_dirs / 'keymaps'
+
+        if (keymap_dir / keymap / 'keymap.c').exists():
+            keymap_path = keymap_dir / keymap / 'keymap.c'
+        if (keymap_dir / keymap / 'keymap.json').exists():
```
[the remark this answers is not in the document]
Comment (skullydazed) [reply lib/python/qmk/keymap.py:126]: This matches the behavior in make. If we want to change that we can but I think that's a change that's out of scope for this PR.

File: lib/python/qmk/keymap.py [hunk 5]
```diff
@@ -95,32 +95,71 @@ def write(keyboard, keymap, layout, layers):
...
     """
     # parse all the rules.mk files for the keyboard
-    rules_mk = qmk.makefile.get_rules_mk(keyboard_name)
+    rules = rules_mk(keyboard)
     names = set()
 
-    if rules_mk:
+    if rules:
         # qmk_firmware/keyboards
         keyboards_dir = Path.cwd() / "keyboards"
```
[the remark this answers is not in the document]
Comment (skullydazed) [reply lib/python/qmk/keymap.py:160]:
```suggestion
        keyboards_dir = Path('keyboards')
```

File: lib/python/qmk/info.py [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+        supported_layouts = rules['LAYOUTS'].strip().split()
+        for layout_name in sorted(layouts):
+            if not layout_name.startswith('LAYOUT_'):
+                continue
+            layout_name = layout_name[7:]
+            if layout_name in supported_layouts:
+                supported_layouts.remove(layout_name)
+
+        if supported_layouts:
+            cli.log.error('%s: Missing layout pp macro for %s' % (keyboard, supported_layouts))
```
[the remark this answers is not in the document]
Comment (skullydazed) [reply lib/python/qmk/info.py:140]:
```suggestion
            cli.log.error('%s: Missing LAYOUT() macro for %s' % (keyboard, ', '.join(supported_layouts)))
```

File: lib/python/qmk/c_parse.py [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+    return files
+
+
+def find_layouts(file):
+    """Returns list of parsed LAYOUT preprocessor macros found in the supplied include file.
+    """
+    file = Path(file)
+    aliases = {}  # Populated with all `#define`s that aren't functions
+    parsed_layouts = {}
+    writing_layout = False
```
[the remark this answers is not in the document]
Comment (skullydazed) [reply lib/python/qmk/c_parse.py:32]: It's probably vestigial. This code came over from qmk_api and has been changed a bunch along the way.

File: lib/python/qmk/path.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Erovia) [reply lib/python/qmk/path.py:7]: Alrighty

File: lib/python/qmk/cli/info.py [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+@cli.argument('-l', '--layouts', action='store_true', help='Render the layouts.')
+@cli.argument('-m', '--matrix', action='store_true', help='Render the layouts with matrix information.')
+@cli.argument('-f', '--format', default='friendly', arg_only=True, help='Format to display the data in (friendly, text, json) (Default: friendly).')
+@cli.subcommand('Keyboard information.')
+@automagic_keyboard
+@automagic_keymap
+def info(cli):
+    """Compile an info.json for a particular keyboard and pretty-print it.
+    """
+    # Determine our keyboard(s)
```
[the remark this answers is not in the document]
Comment (skullydazed) [reply lib/python/qmk/cli/info.py:87]: The bug here is in is_keyboard imo, it should return False in that case.

File: lib/python/qmk/keymap.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Erovia) [reply lib/python/qmk/keymap.py:160]: Right, forgot that.

File: lib/python/qmk/info.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (Erovia) [reply lib/python/qmk/info.py:140]: I'm cool with that.

File: lib/python/qmk/keyboard.py [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+from math import ceil
+from pathlib import Path
+
+from qmk.c_parse import parse_config_h_file
+from qmk.makefile import parse_rules_mk_file
+
+
+def is_keyboard(keyboard_folder):
+    """Returns True if keyboard is a valid keyboard_folder.
+    """
```
Comment (skullydazed) [inline lib/python/qmk/keyboard.py:13]:
```suggestion
    """
    if not keyboard_folder:
        return False
```

File: lib/python/qmk/keyboard.py [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+    keyboard = Path(keyboard)
+    cur_dir = Path('keyboards')
+    rules = parse_rules_mk_file(cur_dir / keyboard / 'rules.mk')
+
+    if 'DEFAULT_FOLDER' in rules:
+        keyboard = Path(rules['DEFAULT_FOLDER'])
+
+    for i, dir in enumerate(keyboard.parts):
+        cur_dir = cur_dir / dir
+        rules = {**rules, **parse_rules_mk_file(cur_dir / 'rules.mk')}
```
[the remark this answers is not in the document]
Comment (skullydazed) [reply lib/python/qmk/keyboard.py:58]:
```suggestion
        rules = parse_rules_mk_file(cur_dir / 'rules.mk', rules)
```

**Comment by skullydazed** [comment]
All feedback should have been addressed and I've rebased on master.

File: lib/python/qmk/cli/list/keymaps.py [hunk 11]
```diff
@@ -13,13 +13,9 @@
...
-    except (FileNotFoundError, PermissionError) as e:
-        cli.echo("{fg_red}%s: %s", cli.config.list_keymaps.keyboard, e)
-    except TypeError:
-        cli.echo("{fg_red}Something went wrong. Did you specify a keyboard?")
+    if not is_keyboard(cli.config.list_keymaps.keyboard):
+        cli.log.error('Keyboard %s does not exist!', cli.config.list_keymaps.keyboard)
+        exit(1)
+
+    for name in qmk.keymap.list_keymaps(cli.config.list_keymaps.keyboard):
+        print(name)
```
Comment (Erovia) [inline lib/python/qmk/cli/list/keymaps.py:21]:
```suggestion
        cli.echo('%s', name)
```

File: lib/python/qmk/keymap.py [hunk 12]
```diff
@@ -130,9 +169,9 @@ def list_keymaps(keyboard_name):
...
 
         # if community layouts are supported, get them
-        if "LAYOUTS" in rules_mk:
-            for layout in rules_mk["LAYOUTS"].split():
-                cl_path = Path.cwd() / "layouts" / "community" / layout
+        if "LAYOUTS" in rules:
+            for layout in rules["LAYOUTS"].split():
+                cl_path = Path('layouts/community') / layout
                 if cl_path.exists():
                     names = names.union([keymap for keymap in cl_path.iterdir() if (cl_path / keymap / "keymap.c").is_file()])
```
Comment (Erovia) [inline lib/python/qmk/keymap.py:176]:
```suggestion
                    names = names.union([keymap.name for keymap in cl_path.iterdir() if (keymap / "keymap.c").is_file() or (keymap / "keymap.json").is_file()])
```

Comment (Erovia) [reply lib/python/qmk/keymap.py:176]: `names = names.union([keymap.name for keymap in keymaps_dir.iterdir() if (keymap / "keymap.c").is_file() or (keymap / "keymap.json").is_file()])` is for Line168, but Github does not allow me to comment on it.

File: lib/python/qmk/tests/test_cli_commands.py [hunk 13]
```diff
@@ -84,3 +110,39 @@ def test_list_keymaps_no_keyboard_found():
     result = check_subcommand('list-keymaps', '-kb', 'asdfghjkl')
     assert result.returncode == 0
```
Comment (Erovia) [inline lib/python/qmk/tests/test_cli_commands.py:111]:
```suggestion
    check_returncode(result, 1)
```

For the other `list-keymap` checks the `assert result.returncode == 0` should be switched to `check_returncode(result)`.

File: lib/python/qmk/tests/test_cli_commands.py [hunk 14]
```diff
@@ -84,3 +110,39 @@ def test_list_keymaps_no_keyboard_found():
...
     assert 'does not exist' in result.stdout
+
+
+def test_info():
+    result = check_subcommand('info', '-kb', 'handwired/onekey/pytest')
+    check_returncode(result)
+    assert 'Keyboard Name: handwired/onekey/pytest' in result.stdout
+    assert 'Processor: STM32F303' in result.stdout
+    assert 'LAYOUT:' not in result.stdout
```
Comment (Erovia) [inline lib/python/qmk/tests/test_cli_commands.py:120]:
```suggestion
    assert 'Layout:' not in result.stdout
```

File: lib/python/qmk/tests/test_cli_commands.py [hunk 15]
```diff
@@ -84,3 +110,39 @@ def test_list_keymaps_no_keyboard_found():
...
+    assert 'k0' not in result.stdout
+
+
+def test_info_keyboard_render():
+    result = check_subcommand('info', '-kb', 'handwired/onekey/pytest', '-l')
+    check_returncode(result)
+    assert 'Keyboard Name: handwired/onekey/pytest' in result.stdout
+    assert 'Processor: STM32F303' in result.stdout
+    assert 'LAYOUT:' in result.stdout
```
Comment (Erovia) [inline lib/python/qmk/tests/test_cli_commands.py:129]:
```suggestion
    assert 'Layout:' in result.stdout
```

File: keyboards/handwired/onekey/pytest/keymaps/default/keymap.json [hunk 16]
```diff
@@ -0,0 +1,9 @@
+{
```
Comment (Erovia) [inline keyboards/handwired/onekey/pytest/keymaps/default/keymap.json:1]: https://github.com/qmk/qmk_firmware/blob/master/keyboards/handwired/onekey/keymaps/default_json/keymap.json

File: lib/python/qmk/cli/list/keymaps.py (hunk 11 above)
Comment (skullydazed) [reply lib/python/qmk/cli/list/keymaps.py:21]: What does this gain us?

Comment (Erovia) [reply lib/python/qmk/cli/list/keymaps.py:21]: Consistency

**Pull request merged**